Add StellarTomlResolver timeout option

robertDurst opened the pull request:
Closes: #158 

## Motivation
Say you are a service that wants to query all asset issuers and are using `StellarSdk.StellarTomlResolver.resolve(uri.host())` to get the TOML information. Unfortunately this sometimes gets caught up and can lag for long periods of time. 

An example url that causes this lag: `xggtoken.com`

## Solution
Add an optional timeout field to the opts param in `StellarSdk.StellarTomlResolver.resolve`. 

Using the [timeout](https://github.com/axios/axios#config-order-of-precedence) config in the axios fetch, we can let sdk users specify a timeout length (in `ms`) to avoid these long TOML resolve lags.

In case users don't want to use the `timeout` config, it is set to 0 by default.
## Example usage:

`await StellarTomlResolver.resolve("xggtoken.com", { timeout: 1000 });`

## Note
I thought about writing a test for this, however to future proof the test suite, I did not want to test against the example I have above in case this domain changes or is taken down. Is this reasoning correct?

bartekn commented:
It looks good however it won't work if stellar.toml file is requested in federation context (`StellarTomlResover` is [called](https://github.com/stellar/js-stellar-sdk/blob/master/src/federation_server.js#L115) in `FederationServer`). To solve this I think we could add `timeout` to `Config` class and use it like `allowHttp`.

> I thought about writing a test for this, however to future proof the test suite, I did not want to test against the example I have above in case this domain changes or is taken down. Is this reasoning correct?

You are correct. You can start a simple HTTP server like [here](https://github.com/stellar/js-stellar-sdk/blob/master/test/unit/federation_server_test.js#L213) with a `setTimeout` that sends a response after a longer time period.

robertDurst commented:
Made the following changes:
1. Added `timeout` to `Config`
2. Added tow unit tests, one where I specify `timeout` in Config, other where I specific in `StellarTomlResolver` opts param

robertDurst commented:
I also see that this `allowHttp` opts param is set in:
* server
* federation server

However, unless I am mistaken, it looks like the `opts` is function specific and since you can now set `timeout` on the `Config` I don't think it is necessary to allow for `opts.timeout`.

bartekn commented on src/config.js:
```diff
@@ -40,6 +53,14 @@ class Config {
+   * Returns the value of `timeout` flag.
+   * @static
+   */
+  static hasTimeout() {
```
Should be `getTimeout`.

bartekn commented on test/unit/stellar_toml_resolver_test.js:
```diff
@@ -106,5 +106,32 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
+    it("rejects after given timeout when global Config.timeout flag is set", function (done) {
+      StellarSdk.Config.setTimeout(1000);
+
+      StellarSdk.StellarTomlResolver.resolve('acme.com')
```
Tests shouldn't make network connections (other to local test server).

bartekn commented:
> However, unless I am mistaken, it looks like the `opts` is function specific and since you can now set `timeout` on the `Config` I don't think it is necessary to allow for `opts.timeout`.

I think it's good to have a global and local settings. For example you may want to change the `timeout` value per request.

We should also add `timeout` to `Server` and `FederationServer`.

robertDurst commented:
Fixed the above:

1. `hasTimeout` --> `getTimeout`
2. Reworked `StellarTomlResolver` test so that it used a local server.

Also looked at where the `opts.allowHttp` was used.

Besides `stellar_toml_resolver.js` it is used in:
* `server.js`:
   * in the constructor to ensure no connection to insecure horizon server.
* `federation_server.js`:
   * in the constructor to ensure no connection to insecure horizon server.
   * for querying StellarTomlResolver via `createForDomain`

So I added it to `federation_server.js` for `createForDomain` and also added a test to `federation_server.js`.

robertDurst commented:
Following my last commit that addressed your feedback, I have a few questions:
* Should I include options (such as allowHttp) even in places where it won't be used (such as in some of the federation_server methods)?
* For `resolveAddress`, `resolveAccountId`, `resolveTransactionId` in the `federation_server` is it best to pass in opts like I do here?
* For call_builder should I pass in `timeout` as an `opts` param like I do or should I just pass in `timeout`?

bartekn commented on src/config.js:
```diff
@@ -32,6 +35,16 @@ class Config {
   }
 
+  /**
+   * Sets `timeout` flag globally. When set to anything besides 0, StellarTOMLResolver will timeout after specified time (ms).
```
Update the comment. It's not `StellarTOMLResolver` only anymore.

bartekn commented on src/federation_server.js:
```diff
@@ -20,6 +20,7 @@ export const FEDERATION_RESPONSE_MAX_SIZE = 100 * 1024;
+ * @param {number} [opts.timeout] - Allow a timeout, default: 0. In ms.
  */
 export class FederationServer {
   constructor(serverURL, domain, opts = {}) {
```
Static method instantiate this class but `opts.timeout` is not persisted anywhere.

bartekn commented on src/federation_server.js:
```diff
@@ -125,43 +128,62 @@ export class FederationServer {
+
+    if (typeof opts.allowHttp !== 'undefined') {
+        allowHttp = opts.allowHttp;
+    }
```
No need for `allowHttp` processing here, it's been checked in `constructor`.

bartekn commented on src/federation_server.js:
```diff
@@ -125,43 +128,62 @@ export class FederationServer {
    * @see <a href="https://www.stellar.org/developers/learn/concepts/federation.html" target="_blank">Federation doc</a>
    * @param {string} transactionId Transaction ID (ex. `3389e9f0f1a65f19736cacf544c2e825313e8447f569233bb8db39aa607c8889`)
+   * @param {object} [opts]
+   * @param {boolean} [opts.allowHttp] - Allow connecting to http servers, default: `false`. This must be set to false in production deployments!
```
`allowHttp` is not needed here as it's checked in `constructor` _per server_.

bartekn commented on src/server.js:
```diff
@@ -32,6 +32,7 @@ export const SUBMIT_TRANSACTION_TIMEOUT = 60*1000;
  * @param {string} serverURL Horizon Server URL (ex. `https://horizon-testnet.stellar.org`).
  * @param {object} [opts]
  * @param {boolean} [opts.allowHttp] - Allow connecting to http servers, default: `false`. This must be set to false in production deployments! You can also use {@link Config} class to set this globally.
+ * @param {number} [opts.timeout] - Allow a timeout, default: 0. In ms.
```
Not passed to `CallBuilder` anywhere.

robertDurst commented:
@bartekn started adding `timeout` to `callbuilder` and realized `callbuilder` was the parent of all the different callbuilders like `TradesCallbuilder` and that furthermore `TradesCallbuilder` was only called from `server`... 

Since that code started to get a bit long and since this PR is focused on solving #158 which only deals with `Federation Server` and `Stellar Toml Resolver` that are seperate from `server` and `callbuilder`, I decided to cut out the extra code for `callbuilder`. 

**So what I now have is just the code for `Federation Server` and `StellarTomlResolver` that directly addresses #158.**

Is this reasoning sound? I think adding `timeout` to callbuilder may be worth a seperate PR.

bartekn commented on src/federation_server.js:
```diff
@@ -161,7 +163,9 @@ export class FederationServer {
 
   _sendRequest(url) {
-    return axios.get(url.toString(), {maxContentLength: FEDERATION_RESPONSE_MAX_SIZE})
+    let timeout = this.timeout | Config.getTimeout();
```
`this.timeout` is not set anywhere.

In general it should work like this:
1. In `constructor` set `this.timeout` to `opts.timeout` or `Config.getTimeout()` (if former not set).
2. `static` method create `FederationServer` instance internally so we should just pass `opts` to class constructor.
3. Finally `_sendRequest` should just use `this.timeout`. No need to check `Config` again.

Also, should be `||`. `|` is bitwise OR.

bartekn commented on test/unit/federation_server_test.js:
```diff
@@ -227,5 +227,98 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
+
+  describe('FederationServer times out when response lags and timeout set', function () {
+    it("resolveAddress times out", function (done) {
+      StellarSdk.Config.setTimeout(100);
```
We should reset it to default value so it doesn't influence other tests. Maybe in `afterEach`? Would be great to have `Config.setDefaults()` or something like this, in case default values change in a future.

bartekn commented on test/unit/federation_server_test.js:
```diff
@@ -227,5 +227,98 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
+        setTimeout(() => {}, 10000);
+      }).listen(4444, () => {
+        StellarSdk.FederationServer
+          .resolve('bob*localhost:4444', {allowHttp: true, timeout: 100})
```
This test shouldn't work because `opts` is not passed to `FederationServer.constructor` (see other comment). It only works because `StellarSdk.Config.setTimeout(100);` is not reset after previous tests.

bartekn commented on test/unit/federation_server_test.js:
```diff
@@ -227,5 +227,98 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
+      let tempServer = http.createServer((req, res) => {
+        setTimeout(() => {}, 10000);
+      }).listen(4444, () => {
+        new StellarSdk.FederationServer('http://localhost:4444/federation', 'stellar.org', {allowHttp: true})
```
The last missing test case is creating a new `FederationServer` with `opts.timeout` set (without calling `Config.setTimeout`).

I think you can create a loop that will test all the methods (`resolveTransactionId`, `resolveAccountId`, ...) with both cases: `Config` in the first iteration and `opts` in the second. Without it we will have a lot of code duplication here.

bartekn commented on test/unit/federation_server_test.js:
```diff
@@ -227,5 +228,71 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
+      // Unable to create temp server in a browser
+      if (typeof window != 'undefined') {
+        return done();
+      }
```
This has to be checked inside `it`s functions. `done` is not defined here. CI tests pass because we do not perform SauceLabs tests in PRs at it could leak our keys.

https://docs.travis-ci.com/user/environment-variables/#Defining-encrypted-variables-in-.travis.yml
> Encrypted environment variables are not available to pull requests from forks due to the security risk of exposing such information to unknown code.

robertDurst replied:
ah, ok had to install firefox for this one

bartekn commented on test/unit/stellar_toml_resolver_test.js:
```diff
@@ -106,5 +106,42 @@ FEDERATION_SERVER="https://api.stellar.org/federation"
     });
+
+    it("rejects after given timeout when global Config.timeout flag is set", function (done) {
+      StellarSdk.Config.setTimeout(1000);
```
Please add `Config.setDefault();` to `afterEach` here too.
